googlecloud-plugins/src/main/java/co/cask/hydrator/plugin/sink/GCSAvroBatchSink.java

```diff
@@ -0,0 +1,122 @@
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.hydrator.plugin.common.StructuredToAvroTransformer;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapreduce.AvroKeyOutputFormat;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapred.JobContext;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * {@link GCSAvroBatchSink} that stores data in avro format to Google Cloud Storage.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name("GCSAvro")
+@Description("Batch sink to write to Google Cloud Storage in Avro format.")
+public class GCSAvroBatchSink extends GCSBatchSink<AvroKey<GenericRecord>, NullWritable> {
+
+  private StructuredToAvroTransformer recordTransformer;
+  private final GCSAvroSinkConfig config;
+
+  public GCSAvroBatchSink(GCSAvroSinkConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    recordTransformer = new StructuredToAvroTransformer(config.schema);
+  }
+
+  @Override
+  protected OutputFormatProvider createOutputFormatProvider(BatchSinkContext context) {
+    return new GCSAvroOutputFormatProvider(config, context);
+  }
+
+  @Override
+  public void transform(StructuredRecord input,
+                        Emitter<KeyValue<AvroKey<GenericRecord>, NullWritable>> emitter) throws Exception {
+    emitter.emit(new KeyValue<>(new AvroKey<>(recordTransformer.transform(input)), NullWritable.get()));
+  }
+
+  /**
+   * Configuration for the GCSAvroSink.
+   */
+  public static class GCSAvroSinkConfig extends GCSSinkConfig {
+
+    @Name("schema")
+    @Description("The Avro schema of the record being written to the sink as a JSON object")
+    private String schema;
+
+    @VisibleForTesting
+    public GCSAvroSinkConfig(String referenceName, String bucketKey, String schema, String projectId,
+                             String serviceKeyFile,
+                             String filesystemProperties, String path) {
+      super(referenceName, bucketKey, projectId, serviceKeyFile,
+            filesystemProperties, path);
+      this.schema = schema;
+    }
+  }
+
+  /**
+   * Output format provider that sets avro output format to be use in MapReduce.
+   */
+  public static class GCSAvroOutputFormatProvider implements OutputFormatProvider {
+
+    private final Map<String, String> conf;
+
+    public GCSAvroOutputFormatProvider(GCSAvroSinkConfig config, BatchSinkContext context) {
+      conf = new HashMap<>();
+      conf.put(JobContext.OUTPUT_KEY_CLASS, AvroKey.class.getName());
+      conf.put("avro.schema.output.key", config.schema);
+      conf.put(FileOutputFormat.OUTDIR,
```

Comment:
This property should be set by the parent class since it will be required for any GCS sink.